torch/utils/_benchmark/utils/valgrind_wrapper/timer_interface.py

```diff
@@ -0,0 +1,448 @@
+        self._validate()
+        baseline_inclusive_stats: Tuple[Tuple[int, str], ...] = ()
+        baseline_exclusive_stats: Tuple[Tuple[int, str], ...] = ()
+        if collect_baseline:
+            cache_key = (number, num_threads)
+            if cache_key not in self._baseline_cache:
+                self._baseline_cache[cache_key] = self._invoke(
+                    stmt="pass", setup="pass", number=number, num_threads=num_threads)
+            baseline_inclusive_stats, baseline_exclusive_stats = \
+                self._baseline_cache[cache_key]
+
+        stmt_inclusive_stats, stmt_exclusive_stats = self._invoke(
+            stmt=stmt,
+            setup=setup,
+            number=number,
+            num_threads=num_threads
+        )
+
+        return CallgrindStats(
+            stmt=stmt,
+            setup=setup,
+            number=number,
+            num_threads=num_threads,
+            built_with_debug_symbols=self._build_type == "RelWithDebInfo",
+            baseline_inclusive_stats=baseline_inclusive_stats,
+            baseline_exclusive_stats=baseline_exclusive_stats,
+            stmt_inclusive_stats=stmt_inclusive_stats,
+            stmt_exclusive_stats=stmt_exclusive_stats,
+        )
+
+    def _invoke(self, stmt: str, setup: str, number: int, num_threads: int):
+        """Core invocation method for Callgrind collection.
+
+        Valgrind operates by effectively replacing the CPU with an emulated
+        version which allows it to instrument any code at the cost of severe
+        performance degradation. This has the practical effect that in order
+        to collect Callgrind statistics, a new process has to be created
+        running under `valgrind`. The steps for this process are:
+
+        1) Create a scratch directory.
+        2) Codegen a run script. (_ValgrindWrapper._construct_script)
+            Inside the run script:
+                * Validate that Python and torch match the parent process
+                * Validate that it is indeed running under valgrind
+                * Execute `setup` and warm up `stmt`
+                * Begin collecting stats
+                * Run the `stmt` loop
+                * Stop collecting stats
+        3) Parse the run results.
+        4) Cleanup the scratch directory.
+        """
+        working_dir = tempfile.mkdtemp()
+        script_file = os.path.join(working_dir, "timer_callgrind.py")
+        callgrind_out = os.path.join(working_dir, "callgrind.out")
+        error_log = os.path.join(working_dir, "error.txt")
+        stat_log = os.path.join(working_dir, "callgrind_stat.txt")
+        stdout_stderr_log = os.path.join(working_dir, "stdout_stderr.log")
+
+        def run(args, **kwargs):
+            # https://thraxil.org/users/anders/posts/2008/03/13/Subprocess-Hanging-PIPE-is-your-enemy/
+            f_stdout_stderr = open(stdout_stderr_log, "wb")
```

Comment:
I don't think this blog post is right; or at least, it is not telling the full story. Here is a test I did on Python 3.6 on Linux 4.16.18:

```
(/home/ezyang/local/pytorch-tmp-env) [ezyang@devvm066.ash0 ~/local/pytorch-tmp] python test2.py 100000
100000
(/home/ezyang/local/pytorch-tmp-env) [ezyang@devvm066.ash0 ~/local/pytorch-tmp] python test2.py 1000000
1000000
(/home/ezyang/local/pytorch-tmp-env) [ezyang@devvm066.ash0 ~/local/pytorch-tmp] python test2.py 10000000
10000000
(/home/ezyang/local/pytorch-tmp-env) [ezyang@devvm066.ash0 ~/local/pytorch-tmp] python test2.py 100000000
100000000
(/home/ezyang/local/pytorch-tmp-env) [ezyang@devvm066.ash0 ~/local/pytorch-tmp] python test2.py 1000000000
1000000000
(/home/ezyang/local/pytorch-tmp-env) [ezyang@devvm066.ash0 ~/local/pytorch-tmp] cat test2.py
from subprocess import Popen, PIPE
import sys
n = sys.argv[1]
p = Popen(('perl', '-e', f"print \"=\" x {n}"), stdout=PIPE, stderr=PIPE)
(out, err) = p.communicate()
print(len(out))
```

Sure, it's guzzling memory, but it's not deadlocking. The most plausible explanation is that this was a bug in Python 2.4 and that was subsequently fixed. I don't think `communicate()` is supposed to deadlock by design in cases where you are willing to fit the entire output in memory. And even if it is, it's not clear to me that code fragments for timing provided by users are going to push us into this regime anyway.

Reply:
I kind of care about knowing if this blog post is right or not, because if it *is* right, we have a lot of subprocess code in PyTorch that needs fixing.

Reply:
I actually was deadlocking on CircleCI until I switched the `callgrind_annotate` step from `PIPE` to file. It may well be platform or version specific. I replicated your script on my machine and also failed to deadlock. (I'd love not to have to do this.)

The cases when a user's code might push this is if it:
A) triggers a lot of noisy warnings.
B) dies, and dumps a ton of debug info. (I have less than fond memories of TF OOM-ing and printing every single gpu allocation AND the C++ stack for every currently running thread.)

In both cases, surfacing the issue is infinitely better than hanging.

Reply:
Do you have the commit for the version in question? :)

Reply:
`ebd453f4f3c987e78ddfd5f833dcfeb35a13a9d8` The timeouts were on `929be913b3bf6d28c5ffaab954eac7024e8e9bbf`.